### DISCUSSION TITLE
How to calculate the house load

### DISCUSSION THREAD

**Discussion by Sammy990** [wlcrs/huawei_solar#315] [Setting up the integration] [answered]
Hi,

I used to collect my house total load all the time, by using a IR sensor and reading the data out of my smart meter.

Now, since my PV has been installed I can't do this anymore, since when PV is producing energy or energy is being supplied by the Luna, the smart meter would say 0W. Is there any way or sensor on the power meter to calculate the load of my house? I just want the plain load, nothing subtracted or cumulated with regards to the PV.

Thanks

**Comment by tarmor1** [comment]
Hi!
Indeed.
This is my approach for it are sensors created as:
```
      - name: "Phases total to Grid"
        state: "{{ max(0,states('sensor.power_meter_phase_a_active_power')|float|round(0))
                 + max(0,states('sensor.power_meter_phase_b_active_power')|float|round(0))
                 + max(0,states('sensor.power_meter_phase_c_active_power')|float|round(0))
                }}"
        device_class: power
        unit_of_measurement: W
        state_class: measurement

      - name: "Phases total from Grid"
        state: "{{ min(0,states('sensor.power_meter_phase_a_active_power')|float|round(0))|abs
                 + min(0,states('sensor.power_meter_phase_b_active_power')|float|round(0))|abs
                 + min(0,states('sensor.power_meter_phase_c_active_power')|float|round(0))|abs
                }}"
        device_class: power
        unit_of_measurement: W
        state_class: measurement


      - name: "Actual Consumption Power"
        state: "{{ (
                     states('sensor.inverter_active_power')|float
                  -  states('sensor.phases_total_to_grid')|float
                  +  states('sensor.phases_total_from_grid')|float
                    )|round(0)
               }}"
        device_class: power
        unit_of_measurement: W
        state_class: measurement
```
This would take then the Solar Total Output, **minus** what I give to the grid (and the difference is what I keep to myself=consumption). And, **added** value that I get from the Grid.
Bear in mind that as 3 Phase inverter, Huawei divides the power received from PV (almost) equally to 3 phases, resulting that phase(s) could be needed extra from grid, while other(s) selling to grid.
Eg if your actual power consumption being P1 200W, P2 300W and P3 1000W and your solar produces 1800W (600W to each phase), then P1 goes to +400W, P2 +300W and P3 -400W.
Minus being buy from grid, plus as sell to grid.

**Reply by Sammy990** [reply]
Isn't there a sensor from the power meter already? I mean I just want to know what or how much electricity load is requested from all household devices like refrigerator etc. at the moment. I thought the power meter should measure the house load and max have a published value for the requested load. At this point I don't want to know how much is coming from PV and/or Battery. Just the plain load from the house itself.

**Reply by JacoboSanchez** [reply]
As I understand it the devices (and so the integration) usually gives you information about how much comes in and how much gets out of the device. A meter located between the grid and your house would tell you if there is power being sold or you are buying it. If you have a battery that is powering your house, then the meter does not know anything but that there is no power coming from the grid and you are not sending any power to the grid so the value returned by modbus is 0.

So if you want to compute how much power your house is getting you have two options (I may be wrong):

- To get values on how much power is coming out and in from each device (inverter, panels, battery, grid) using data obtained by modbus interface with this integration and do the maths on what is finally used by your house.
- To have a meter with the clamp attached to the central wire powering your house and read the data from it (for example a Shelly EM device)

**Comment by JacoboSanchez** [comment, marked as answer]
I have posted my configuration at some point but will post the relevant sensors added to try to identify the flows of power from one entity to other being house_consumption the sum of battery_to_house, grid_to_house and generation_to_house. There may be errors but at least may provide some ideas on how to compute what you want with the data provided by the inverter. Also this relies on base sensors that have different names than the ones provided currently from the integration (for example my template storage_charge_discharge can be directly replaced by the battery_charge_discharge_power created by the integration... this is to some names normalizations I did due to early stage integration changes)

```
      battery_charge:
        friendly_name: "Battery charge"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set u = states('sensor.storage_charge_discharge') | int %}
          {% if u > 0 %}
              {{ (u) | int }}
          {% else %}
              {{ (0) | int }}
          {% endif %}
        availability_template: >-
          {{ (states('sensor.storage_charge_discharge') not in [ 'unavailable', 'none', 'unknown' ] ) }}


      generation_to_battery:
        friendly_name: "Generation to battery"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set b = states('sensor.battery_charge') | int %}
          {% set ap = states('sensor.input_power') | int %}
          {% if b > 0 and ap > 0 and b < ap %}
              {{ (b) | int }}
          {% elif b > 0 and ap > 0 and b > ap %}
              {{ (ap) | int }}
          {% else %}
              {{ (0) | int }}
          {% endif %}
        availability_template: >-
          {{ (states('sensor.battery_charge') not in [ 'unavailable', 'none', 'unknown' ] )
              and (states('sensor.input_power') not in [ 'unavailable', 'none', 'unknown' ]) }}
      
      grid_to_battery:
          friendly_name: "Grid to battery"
          unit_of_measurement: "W"
          device_class: power
          value_template: >-
            {% set ip = states('sensor.active_power') | int %}
            {% set cdp = states('sensor.storage_charge_discharge') | int %}
            {% if ip < 0 and cdp > 0 %}
                {{ (-1 * ip) | int }}
            {% else %}
                {{ (0) }}
            {% endif %}
          availability_template: >-
            {{ (states('sensor.active_power') not in [ 'unavailable', 'none', 'unknown' ] )
                and (states('sensor.storage_charge_discharge') not in [ 'unavailable', 'none', 'unknown' ]) }}


      battery_to_house:
        friendly_name: "Battery to house"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set u = states('sensor.storage_charge_discharge') | int %}
          {% if u < 0 %}
              {{ (- u) | int }}
          {% else %}
              {{ (0) | int }}
          {% endif %}
        availability_template: >-
          {{ (states('sensor.storage_charge_discharge') not in [ 'unavailable', 'none', 'unknown' ] ) }}
  
      grid_to_house:
        friendly_name: "Grid to house"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set g_c = states('sensor.grid_consumption') | int %}
          {% set g_t_b = states('sensor.grid_to_battery') | int %}
          {{ (g_c-g_t_b) | int }}
        availability_template: >-
          {{ (states('sensor.grid_consumption') not in [ 'unavailable', 'none', 'unknown' ] )
              and (states('sensor.grid_to_battery') not in [ 'unavailable', 'none', 'unknown' ]) }}   

      generation_to_house:
        friendly_name: "Generation to house"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set pv = states('sensor.input_power') | int %}
          {% set g_t_b = states('sensor.generation_to_battery') | int %}
          {% set g_t_gr = states('sensor.generation_to_grid') | int %}
          {% if (pv - g_t_b - g_t_gr) > 0 %}
              {{ (pv - g_t_b - g_t_gr) | int }}
          {% else %}
              {{ (0) }}
          {% endif %}
        availability_template: >-
          {{ (states('sensor.input_power') not in [ 'unavailable', 'none', 'unknown' ] )
              and (states('sensor.generation_to_battery') not in [ 'unavailable', 'none', 'unknown' ]) 
              and (states('sensor.generation_to_grid') not in [ 'unavailable', 'none', 'unknown' ]) }}      
      
      house_consumption:
        friendly_name: "House consumption"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set t = states('sensor.battery_to_house') | int %}
          {% set u = states('sensor.grid_to_house') | int %}
          {% set g = states('sensor.generation_to_house') | int %}
          {{ (t + u + g) }}
        availability_template: >-
          {{ (states('sensor.battery_to_house') not in [ 'unavailable', 'none', 'unknown' ] )
              and (states('sensor.grid_to_house') not in [ 'unavailable', 'none', 'unknown' ]) 
              and (states('sensor.generation_to_house') not in [ 'unavailable', 'none', 'unknown' ]) }}
```

**Reply by PStenbumling** [reply]
sensor.input_power and sensor.active_power is that from the inverter?  e.g. sensor.inverter_input_power and sensor.inverter_active_power

**Reply by JacoboSanchez** [reply]
I have a more complex configuration with suffixes in sensor template names so I tried to paste only relevant information as base sensor names may vary due to me having two configurations of the integration but will try to paste the base templates also if you find them useful.

As mentioned above storage_charge_discharge is just a rename of battery_charge_discharge_power:
 ```
     storage_charge_discharge:
        friendly_name: "Storage Charge-Discharge"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set s = states('sensor.battery_charge_discharge_power') | int %}
          {{ (s) }}
        availability_template: >-
          {{ (states('sensor.battery_charge_discharge_power') not in [ 'unavailable', 'none', 'unknown' ] )}}
```

Input power is a renaming of inverter_input_power:
```
input_power:
        friendly_name: "Input Power"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set s = states('sensor.inverter_input_power') | int %}
          {{ (s) }}
        availability_template: >-
          {{ (states('sensor.inverter_input_power') not in [ 'unavailable', 'none', 'unknown' ] )}}
```

Active power is a renaming of inverter_active_power:
```
      active_power:
        friendly_name: "Active Power"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set s = states('sensor.inverter_active_power') | int %}
          {{ (s) }}
        availability_template: >-
          {{ (states('sensor.inverter_active_power') not in [ 'unavailable', 'none', 'unknown' ] )}}
```


You don't need to use such renamings as they will not add any value for you. In my scenario they are more useful because I am mapping (in the samples above I have deleted the suffix) names from integration to location-based names:
```
sensor.inverter_active_power => sensor_active_power_locationA
sensor.inverter_active_power_1 => sensor_active_power_locationB
```
So I can use the names based on location and not by automatic `_1` renamings from the duplicated integration configurations

**Reply by PStenbumling** [reply]
Is sensor.grid_consumption the sensor.power_meter_consumption?

**Reply by PStenbumling** [reply]
Im trying to calculate some kind of revenue.
its different base on where the input goes.

**Reply by JacoboSanchez** [reply]
I missed the definition of that one. Grid consumption is the value of the `sensor.power_meter_active_power` when its value is negative (negative means buying power, positive selling power). Usually grid consumption is used by the house, but there are scenarios where you can force the battery charge from the grid so you will have grid consumption going also to the battery. 

This shows you how I mapped sensor values of incoming/outgoing power to somehow deduce power lines (mainly to generate a power flow diagram card). 

You must identify what do you need to know and figure out how to compute it with the measures provided by the integration. 

Usually costs/revenues  are not computed using power but energy (Wh). All of this sensors return power (W). If that is energy what you are looking for none of this sensors would help you directly, there are other sensor providing energy but they are less than the power ones (look for daily or total ones)

```
grid_consumption:
        friendly_name: "Grid consumption"
        unit_of_measurement: "W"
        device_class: power
        value_template: >-
          {% set u = states('sensor.power_meter_active_power') | int %}
          {% if u < 0 %}
              {{ (- u) | int }}
          {% else %}
              {{ (0) | int }}
          {% endif %}
        availability_template: >-
          {{ (states('sensor.power_meter_active_power') not in [ 'unavailable', 'none', 'unknown' ] )}}
```

There are a more detailed description of my configuration along with other versions from other users in [this discussion from tesla-sstyle-solar-power-card](https://github.com/reptilex/tesla-style-solar-power-card/issues/95#issuecomment-1066183727)

**Reply by PStenbumling** [reply]
Big thanks!
Yeah, i needed to split the battery charge/discgarge since i dont want to handle negative values. 
feels like a good start neitherless :)

**Reply by PStenbumling** [reply]
Im thinking a bit about the Battery benefit. 
im not using grid to battery or battery to grid

So when my batteries are charging i have a "over" production. 
is my calculation below correct then? 


```
       - name: Solar Benefit
        unique_id: solar_benefit
        unit_of_measurement: "SEK/h"
        state: >-
          {{states('sensor.generation_to_house')|float(0) * states('sensor.nordpool_buy_total_sek_kwh')|float(0)/1000
          + states('sensor.battery_to_house')|float(0) * states('sensor.nordpool_buy_total_sek_kwh')|float(0)/1000
          + states('sensor.generation_to_grid')|float(0) * states('sensor.nordpool_sell_total_sek_kwh')|float(0)/1000}}
```
          
```
      - name: Battery Benefit
        unit_of_measurement: SEK/h
        state: >
          {{states('sensor.generation_to_battery')|float(0) * states('sensor.nordpool_buy_total_sek_kwh')|float(0)/1000
          + states('sensor.battery_to_house')|float(0) * states('sensor.nordpool_buy_total_sek_kwh')|float(0)/1000}}
```

**Reply by a13xde** [reply]
@JacoboSanchez Thank you for your detailed description, but there is still one unclear sensor: "generation_to_grid"
Can you please help out to complete your config with this sensor ?
Thank you !

**Reply by etamtlosz** [reply]
@a13xde you can find it here [#95](https://github.com/reptilex/tesla-style-solar-power-card/issues/95)

**Reply by JacoboSanchez** [reply]
> @a13xde you can find it here [#95](https://github.com/reptilex/tesla-style-solar-power-card/issues/95)

Yes, I have even posted that link as the final sentence on my comment in case more detail was needed... I wonder why I took the time to describe my configuration and search for the discussion to add as reference if people won't give the effort to even read it to the end

**Reply by a13xde** [reply]
I'm sorry, I really missed this link, thank you for the hint.
And also thank you for your examples on how to work with the sensors from this integration, it really helped me a lot
